File: hoffman-agents-node/src/core/strange-loop.js
const SELF_TOKEN = 'I';

const PREDICATE_TOKENS = new Set([
  'remember', 'notice', 'think', 'see', 'feel', 'know',
  'cross', 'arrive', 'return', 'enter', 'leave',
  'expect', 'predict', 'observe', 'experience', 'am',
]);

function computeSelfReferenceScore(sequence) {
  const nI = sequence.filter(t => t === SELF_TOKEN).length;

  if (nI === 0) return 0;
  if (nI === 1) return 0.5;

  const positions = sequence.map((t, i) => t === SELF_TOKEN ? i : -1).filter(i => i >= 0);
  let validPairs = 0;

  for (let idx = 0; idx < positions.length; idx++) {
    for (let jdx = idx + 1; jdx < positions.length; jdx++) {
      const i = positions[idx], j = positions[jdx];
      const between = sequence.slice(i + 1, j);
      const hasPredicate = between.some(t => PREDICATE_TOKENS.has(t));
      if (hasPredicate) validPairs++;
    }
  }

  if (validPairs === 0) return 0.5;
  const maxDepth = positions.length;
  const score = 0.5 * maxDepth * (validPairs / (nI * (nI - 1) / 2));
  return Math.min(score, 2.0);
}

function populationReferenceScore(sequences) {
  if (!sequences || sequences.length === 0) return 0;
  return sequences.reduce((s, seq) => s + computeSelfReferenceScore(seq), 0) / sequences.length;
}

const strangeLoopScore = computeSelfReferenceScore;
const populationLoopScore = populationReferenceScore;

function firstDepthNGeneration(loopHistory, threshold) {
  for (let gen = 0; gen < loopHistory.length; gen++) {
    if (loopHistory[gen] >= threshold) return gen;
  }
  return null;
}

module.exports = {
  computeSelfReferenceScore,
  populationReferenceScore,
  strangeLoopScore,
  populationLoopScore,
  firstDepthNGeneration,
  SELF_TOKEN,
};

File: hoffman-agents-python/src/conscious_agent/core/strange_loop.py
from __future__ import annotations

from collections.abc import Sequence

SELF_TOKEN = "I"

_PREDICATE_TOKENS = frozenset({
    "remember", "notice", "think", "see", "feel", "know",
    "cross", "arrive", "return", "enter", "leave",
    "expect", "predict", "observe", "experience", "am",
})


def compute_self_reference_score(sequence: Sequence[str]) -> float:
    n_i = sum(1 for t in sequence if t == SELF_TOKEN)

    if n_i == 0:
        return 0.0
    if n_i == 1:
        return 0.5

    positions = [i for i, t in enumerate(sequence) if t == SELF_TOKEN]
    valid_pairs = 0

    for idx, i in enumerate(positions):
        for j in positions[idx + 1:]:
            between = sequence[i + 1:j]
            has_predicate = any(t in _PREDICATE_TOKENS for t in between)
            if has_predicate:
                valid_pairs += 1

    if valid_pairs == 0:
        return 0.5

    max_depth = len(positions)
    score = 0.5 * max_depth * (valid_pairs / (n_i * (n_i - 1) / 2))

    return min(score, 2.0)


def population_reference_score(sequences: Sequence[Sequence[str]]) -> float:
    if not sequences:
        return 0.0
    return sum(compute_self_reference_score(s) for s in sequences) / len(sequences)


strange_loop_score = compute_self_reference_score
population_loop_score = population_reference_score


def first_depth_n_generation(
    loop_history: Sequence[float],
    threshold: float,
) -> int | None:
    for gen, score in enumerate(loop_history):
        if score >= threshold:
            return gen
    return None
